corehq/motech/repeaters/models.py

```diff
@@ -970,6 +970,27 @@ class Meta:
 
     payload_generator_classes = (DataSourcePayloadGenerator,)
 
+    def register(self, payload, fire_synchronously=False):
+        """
+        Merge DataSourceUpdates of waiting repeat records.
+
+        To avoid a race condition where a waiting record is sent while
```

Comment:
Would this run into performance issues?

For every repeat record getting registered we are updating all that are already present for the data source. 
I assume it's mostly going to be just a one more record that already exists?
So, we are canceling the record that we consider being sent in the race condition? it is possible that we will cancel it while it is being fired?

And this is all happening as a part of a signal call from a form submission/case update.

I wonder if these operations are too complicated to be done here.

Reply:
> it is possible that we will cancel it while it is being fired?

Yes. This is fine. It is already possible for a DataSourceRepeater to update the same rows more than once.

> I wonder if these operations are too complicated to be done here.

* It is not much more complicated than the current process for building and sending a payload.
* It saves a gigantic amount of time compared to sending a payload for every form submission.

@gherceg is more familiar with the load and resources used by Repeaters than I am. @gherceg would you be able to weigh in on this change? If you think this operation is too heavy up-front, I'd be keen to hear your thoughts on alternative approaches.

Reply:
> I assume it's mostly going to be just a one more record that already exists?

Can y'all elaborate on why it is mostly going to just be one more record than already exists?

Reply:
@gherceg because of this scenario:

1. Repeat record 1 is registered. `register()` calls `repeat_record.attempt_forward_now()`, and the send attempt fails.
2. Repeat record 2 is registered. It finds repeat record 1, and cancels it.
3. Repeat record 3 is registered. It only finds repeat record 2, because repeat record 2 cancelled repeat record 1. Repeat record 3 cancels repeat record 2.
4. ... etc.

Reply:
> updating the existing repeat record

Do you have a way of avoiding the race condition I described [in the thread above](https://github.com/dimagi/commcare-hq/pull/36444#discussion_r2136664613)?

Reply:
Ah sorry missed that. Yeah that does seem like a dealbreaker for updating existing records. As far as this thread/questions about performance of cancelling existing records, I don't have any more followup questions.